### PR TITLE
Add missing annotation support for service objects and improve listener

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Add back missing annotation support for services. 

And allow listener's `attach` method to control what to accept as service path.
Second param of listener.attach method can be any combination of `string[], string, ()`.

Fixes: https://github.com/ballerina-platform/ballerina-lang/issues/26885, https://github.com/ballerina-platform/ballerina-lang/issues/27348, and https://github.com/ballerina-platform/ballerina-lang/issues/26838

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/AnnotationUtils.java [hunk 1]
```diff
@@ -43,12 +43,17 @@
...
      */
-    public static void processAnnotations(MapValue globalAnnotMap, Type bType) {
+    public static void processAnnotations(MapValue globalAnnotMap, Type bType, Strand strand) {
         if (!(bType instanceof BAnnotatableType)) {
             return;
         }
 
         BAnnotatableType type = (BAnnotatableType) bType;
+        if (type.getTag() == TypeTags.SERVICE_TAG) {
+            processServiceAnnotations(globalAnnotMap, (BServiceType) bType, strand);
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/AnnotationUtils.java:53]: `processServiceAnnotations` used to be codegen-ed separately for services to specifically recompute the annotations in services that are created within a function (via https://github.com/ballerina-platform/ballerina-lang/blob/v2.0.0-Preview7/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java#L1625).

To continue to support the same, IMO we need to do something similar.

The second annotation access in the following sample used to return `{i: 202}`, but now returns `{i: 101}`.

```ballerina
annotation record {int i;} x on service;

int globalInt = 101;

public function getService() returns service object {} {
    return @x {i: globalInt} service object {
        
    };
}

public function main() {
    typedesc<service object {}> t = typeof getService();
    var v = t.@x; // {"i":101}

    globalInt = 202;
    t = typeof getService();
    v = t.@x; // used to be {"i":202}, now {"i":101}
}
```

Will we be fixing this anytime soon given that it's a breaking change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -3798,18 +3796,25 @@ private boolean checkAttachMethod(BAttachedFunction func) {
...
+
+        private boolean isServiceObject(BType type) {
+            if (type.tag != TypeTags.OBJECT) {
+                return false;
+            }
+
+            if (!types.isServiceObject((BObjectTypeSymbol) type.tsymbol)) {
+                return false;
+            }
+            return true;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:3817]:
```suggestion
            return types.isServiceObject((BObjectTypeSymbol) type.tsymbol);
```

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 3]
```diff
@@ -775,6 +775,22 @@ error.invalid.listener.var=\
...
   invalid listener attachment
 
+error.service.absolute.path.or.literal.required.by.listener=\
+  service absolute path or literal is required by listener
+
+error.service.path.literal.is.not.supported.by.listener=\
+  service path literal is not supported by listener
+
+error.service.absolute.path.is.not.supported.by.listener=\
+  service absolute path is not supported by listener
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:785]: Will these be supported in the future?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 4]
```diff
@@ -775,6 +775,22 @@ error.invalid.listener.var=\
...
+
+error.service.path.literal.required.by.listener=\
+  service path literal is required by the listener
+
+error.service.absolute.path.required.by.listener=\
+  service absolute path is required by the listener
+
+
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:793]: Extra new line.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/AnnotationUtils.java (hunk 1 above)
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/AnnotationUtils.java:53]: Could this be this issue, I mean this behavior https://github.com/ballerina-platform/ballerina-lang/issues/26838

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 3 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/resources/compiler.properties:785]: No this is the expected behavior,

if listener attach method is `attach(service object s, string? path)`, then it's an error to have following:
```
service \abs\path on new http:Listener(4000) {

}
```

It should be  either `service "name-literal" on lsn { }` or `service on lsn { }`

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BServiceType.java [hunk 5]
```diff
@@ -99,4 +93,8 @@ public void setResourceFunctions(ResourceFunctionType[] resourceFunctions) {
     public int getTag() {
         return TypeTags.SERVICE_TAG;
     }
+
+    public boolean isAnnotationsAvailable() {
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BServiceType.java:97]: will update

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/AnnotationUtils.java (hunk 1 above)
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/AnnotationUtils.java:53]: Issue to re-visit this annot impl: https://github.com/ballerina-platform/ballerina-lang/issues/27348

I think we will have to consider object constructors for the above issue.

**Review by hasithaa — APPROVED** [review]

**Pull request merged**